src/interpreter/s2s_forward_mode_ad.jl

```diff
@@ -0,0 +1,147 @@
+    try
+        # If we've already derived the OpaqueClosures and info, do not re-derive, just
+        # create a copy and pass in new shared data.
+        oc_cache_key = ClosureCacheKey(interp.world, (sig_or_mi, debug_mode))
+        if haskey(interp.oc_cache, oc_cache_key)
+            return _copy(interp.oc_cache[oc_cache_key])
+        else
+            # Derive forward-pass IR, and shove in a `MistyClosure`.
+            forward_ir = generate_forward_ir(interp, sig_or_mi; debug_mode)
+            fwd_oc = MistyClosure(forward_ir; do_compile=true)
+            raw_rule = DerivedFRule(fwd_oc)
+            rule = debug_mode ? DebugFRule(raw_rule) : raw_rule
+            interp.oc_cache[oc_cache_key] = rule
+            return rule
+        end
+    catch e
+        rethrow(e)
+    finally
+        unlock(MOONCAKE_INFERENCE_LOCK)
+    end
+end
+
+function generate_forward_ir(
+    interp::MooncakeInterpreter,
+    sig_or_mi;
+    debug_mode=false,
+    do_inline=true,
+)
+    # Reset id count. This ensures that the IDs generated are the same each time this
+    # function runs.
+    seed_id!()
+
+    # Grab code associated to the primal.
+    primal_ir, _ = lookup_ir(interp, sig_or_mi)
+
+    # Normalise the IR.
+    isva, spnames = is_vararg_and_sparam_names(sig_or_mi)
+    ir = normalise!(primal_ir, spnames)
+
+    fwd_ir = dualize_ir(ir)
+    opt_fwd_ir = optimise_ir!(fwd_ir; do_inline)
+    return opt_fwd_ir
+end
+
+function dualize_ir(ir::IRCode)
+    new_stmts_stmt = map(make_fwd_ad_stmt, ir.stmts.stmt)
+    new_stmts_type = map(dual_type, ir.stmts.type)
+    new_stmts_info = ir.stmts.info
+    new_stmts_line = ir.stmts.line
+    new_stmts_flag = ir.stmts.flag
+    new_stmts = CC.InstructionStream(
+        new_stmts_stmt,
+        new_stmts_type,
+        new_stmts_info,
+        new_stmts_line,
+        new_stmts_flag,
+    )
+    new_cfg = ir.cfg
+    new_linetable = ir.linetable
+    rule_type = Any
+    new_argtypes = convert(Vector{Any}, vcat(rule_type, map(make_fwd_argtype, ir.argtypes)))
+    new_meta = ir.meta
+    new_sptypes = ir.sptypes
+    return IRCode(new_stmts, new_cfg, new_linetable, new_argtypes, new_meta, new_sptypes)
+end
+
+make_fwd_argtype(::Type{P}) where {P} = dual_type(P)
+make_fwd_argtype(c::Core.Const) = Dual  # TODO: refine to type of const
+
+function make_fwd_ad_stmt(stmt::Expr)
+    interp = get_interpreter()  # TODO: pass it around
+    C = context_type(interp)
+    if isexpr(stmt, :invoke) || isexpr(stmt, :call)
+        mi = stmt.args[1]::Core.MethodInstance
+        sig = mi.specTypes
+        if is_primitive(C, sig)
+            shifted_args = map(stmt.args) do a
+                if a isa Core.Argument
+                    Core.Argument(a.n + 1)
+                else
+                    a
+                end
+            end
+            new_stmt = Expr(
+                :call,
+                :($frule!!),
+                stmt.args[2],
+                shifted_args[3:end]...
```

Comment:
Could you run me through what's going on here?

Reply:
> [!WARNING]
> I'm still finding my footing on metaprogramming, so anything involving expressions is to be taken with a grain of salt.

This is based on https://compintell.github.io/Mooncake.jl/previews/PR386/developer_documentation/forwards_mode_design/#Statement-Transformation. For a primitive signature, we know that `frule!!` has an appropriate method, so we just replace the call
```julia
f(x1, x2, x3)
```
with
```julia
frule!!(f, x1, x2, x3)
```
In the `:invoke` expression, `args[1]` is the method instance, `args[2]` is the function object and `args[3:end]` are the actual arguments. This could have just been `shifted_args[2:end]...`. I interpolated `$frule!!` becaue otherwise Julia seemed to look for that symbol in `Base.Experimental` (no clue why).

I'm still not super clear on what it means for an expression to contain things that are generic Julia objects. The [docs page on metaprogramming](https://docs.julialang.org/en/v1/manual/metaprogramming/#Quoting) contains this very obscure sentence:

> Expressions provided by the parser generally only have symbols, other expressions, and literal values as their args, whereas expressions constructed by Julia code can have arbitrary run-time values without literal forms as args.

What happens to these expression components during evaluation? They are just left as-is, unlike symbols and other expressions?

Reply:
Ahhh I see - this is just meant to handle `invoke`. So `shifted_args[2:end]` are all of the arguments after the `MethodInstance` -- understood.

So I think that you really do have to splat `shifted_args[2:end]` here, rather than using `args[2]` and `shifted_args[3:end]`, because what happens if `args[2]` is itself an `Argument`? i.e. in the general case, it might be that we're calling a function which has been passed in as an argument to the method we're deriving a rule from.

> What happens to these expression components during evaluation? They are just left as-is, unlike symbols and other expressions?

Hmmm well in the `IRCode` context, I'm really confident that we're not permitted to use `Expr`s (other than special ones like `Expr(:boundscheck)`) in `:call` expressions. When I've got a specific function that I want to make use of, I think I've just put the function object itself in the IR. i.e. literally just writing `Expr(:call, frule!!, ...)`. It's somewhat obscured by some layers of abstraction in the method of `make_ad_stmts!(::Expr, ...)` sadly, but if you work backwards from https://github.com/compintell/Mooncake.jl/blob/5397d284e89f08afbb22cce32483c1ada6baec10/src/interpreter/s2s_reverse_mode_ad.jl#L661 you can kind of see how the `rrule!!` object winds up directly inserted in the `:call` expression used on the forwards-pass of reverse-mode.